src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts

```diff
@@ -6,9 +6,63 @@ export module DotNet {
   export type JsonReviver = ((key: any, value: any) => any);
   const jsonRevivers: JsonReviver[] = [];
 
+  class JSObject {
+    public static readonly ID_KEY = '__jsObjectId';
+
+    _cachedFunctions: Map<string, Function>;
+
+    constructor(private _jsObject: any, _id: number)
+    {
+      this._cachedFunctions = new Map<string, Function>();
+      this._jsObject[JSObject.ID_KEY] = _id;
```

Comment:
I don't think we should be mutating the object instance that the developer has supplied to us. That could cause all kinds of side-effects.

Presumably the reason you were doing this is so that we have a single `JSObject` wrapper for each instance. If so:

 * Is that important? Would something bad happen if we created separate wrapper instances each time?
    * Is there a risk that the current design gives rise to a bug whereby code A requests some object, then code B requests the same object for unrelated reasons, then code A disposes it, then code B tries to use it but gets an error because it was disposed by unrelated code? I haven't read all the code paths thoroughly enough to evaluate this for sure, but it seems safer if each usage produces an independent wrapper that can be disposed independently.
 * If we do have to get a single wrapper per instance, then instead of writing the `JSObject.ID_KEY` to it, consider having a `Map<any, number>` in which you store all the tracked instances and their corresponding IDs. Arbitrary objects can be keys in a `Map`.

Reply:
I've read a bit more of the code now and I *think* the issue with disposal does apply. Is that right?

If so I think we should avoid trying to deduplicate `JSObject` instances when an invoked JS function supplies its return value, and have separate wrappers for each returned value.

Perhaps the reason for the original design was that you wanted to be able to do equality checks on the .NET side and see whether two `JSObjectReference` values represent the same JS object instance. If so I appreciate the benefit of that, but the problems with unrelated code disposing things you're still trying to use seem too problematic and would outweigh this benefit.

If we did want the .NET-side equality checking to work, I think we'd need to do something more complicated, whereby we have two identifiers for each `JSObject`: one that represents the wrapper instance, and each wrapper can be disposed independently, and another one that represents the underlying instance. We *could* then override `Equals` on `JSObjectReference` so that two instances would equal each other if the underlying instance ID matches, however that's pretty awkward because it's weird in .NET for `a.Equals(b)` to be true when `a` and `b` are not reference-equal. So perhaps best would be having a separate method on `JSObjectReference` like `a.WrapsSameInstanceAs(b)`. If we did that, it could be added any time in the future non-breakingly, so wouldn't need to be implemented right now.

Reply:
Great suggestions, I'll change my implementation to avoid mutating the object instance.

Reply:
I've gone ahead and implemented this change. I believe the disposal scenario you described is a real scenario, so I think this was a good change to make. I agree that it's probably not necessary to implement the .NET-side equality checking, at least for now.